gapic/src/main/com/google/gapid/util/Paths.java

```diff
@@ -248,82 +272,107 @@ public static int compare(Path.Command a, Path.Command b) {
     return (a.getIndicesCount() == b.getIndicesCount()) ? 0 : -1;
   }
 
-  public static Path.State findState(Path.Any path) {
+  public static boolean contains(Path.Any path, Predicate<Path.Any> predicate) {
+    return find(path, predicate) != null;
+  }
+
+    public static Path.Any find(Path.Any path, Predicate<Path.Any> predicate) {
+    for (Path.Any p = path; p != null; p = parentOf(p)) {
+      if (predicate.test(p)) {
+        return p;
+      }
+    }
+    return null;
+  }
+
+  public static Path.Any parentOf(Path.Any path) {
```

Comment:
This is incomplete and I don't think it's necessary. This is trying to future-proof something without actually making it future-proof. Please undo this and just add the global state case.

Reply:
> This is trying to future-proof something without actually making it future-proof

I'd say it is more than just future-proofing. You had an ad-hoc collection of find functions that would need to be duplicated for `findGlobalState`. Yes, what I have changed to is incomplete, but it's as incomplete as your old `findState`. I'd rather implement the full set than roll back to a bunch of random functions. This approach mirrors the go-side, where parent-traversal is used for numerous cases.

Reply:
>  but it's as incomplete as your old `findState`

I disagree with that assertion. `findState` was incomplete in the sense that the state path was the only path it could find, but it correctly handled every kind of path. `parentOf` pretends to be able to find the parent of any path, but it does not.

I know this is how it's done in go, and it works OK there due to the implicit interfaces. However, in Java the protos are immutable and the constant toBuilder().build() to wrap it in `Path.Any` for it to then just be unwrapped is a lot of overhead for something that is not complete and only used to determine if a path has a state root.

I propose, for now, we simply change `findState` to `isStatePath`, returning a boolean, since that's all that it's actually used for. This makes the change to handle the new global state root very simple. We then look at how we can do the parent relationship better. I would prefer if it was part of the proto.